File: DayFive.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_LINES 1000
#define MAX_DIGITS 20

char file[MAX_LINES][MAX_DIGITS];
int fileLength = 0;

long FromSeedToLocation(long seed);

int compare(const void *a, const void *b);

int main() {
    FILE *filePtr = fopen("input5", "r");

    if (filePtr == NULL) {
        perror("Error opening file");
        return 1;
    }

    while (fileLength < MAX_LINES && fgets(file[fileLength], MAX_DIGITS, filePtr) != NULL) {
        fileLength++;
    }

    fclose(filePtr);

    long locations[MAX_LINES];
    char *token = strtok(file[0], " ,\t\n");

    int index = 0;
    while (token != NULL && index < MAX_LINES) {
        locations[index++] = atol(token);
        token = strtok(NULL, " ,\t\n");
    }

    for (int i = 0; i < fileLength; i++) {
        locations[i] = FromSeedToLocation(locations[i]);
    }

    qsort(locations, fileLength, sizeof(long), compare);

    printf("Part 1: %ld\n", locations[0]);

    return 0;
}

long FromSeedToLocation(long seed) {
    for (int i = 3; i < fileLength; i++) {
        char *token;
        char *line = file[i];
        bool reachedEOF = i >= fileLength - 1;

        token = strtok(line, " ,\t\n");
        long destinationStart = atol(token);

        token = strtok(NULL, " ,\t\n");
        long sourceStart = atol(token);

        token = strtok(NULL, " ,\t\n");
        long range = atol(token);

        while (token != NULL && !reachedEOF) {
            if (seed == sourceStart) {
                seed = destinationStart + (seed - sourceStart);

                while (1) {
                    i++;
                    if (i >= fileLength - 1) {
                        break;
                    }
                }

                break;
            }

            if (i + 1 == fileLength) {
                reachedEOF = 1;
                break;
            }

            i++;

            char *nextLine = file[i];

            if (nextLine == NULL) {
                break;
            }

            token = strtok(nextLine, " ,\t\n");
            if (token == NULL) {
                break;
            }

            destinationStart = atol(token);

            token = strtok(NULL, " ,\t\n");
            if (token == NULL) {
                break;
            }

            sourceStart = atol(token);

            token = strtok(NULL, " ,\t\n");
            if (token == NULL) {
                break;
            }

            range = atol(token);
        }

        if (reachedEOF) {
            return seed;
        }
    }

    return seed;
}

int compare(const void *a, const void *b) {
    return (*(long *)a - *(long *)b);
}

File: DayFive.java
import java.nio.file.*;
import java.util.Arrays;
import java.util.List;
import java.util.stream.IntStream;
import java.util.stream.LongStream;

public class DayFive {
    static List<String> file;
    static int fileLength;
    public static void main(String[] args) {
        try {
            file = Files.readAllLines(Path.of("input5"));
            
        } catch (Exception e) {
            System.out.println(e);
        }
        
        fileLength = file.size();
        long[] locations = Arrays.stream(file.get(0).split("\\D+"))
            .filter(s -> !s.isBlank())
            .mapToLong(Long::parseLong).toArray();
            //.map(e -> LongStream.range(e, e+1))
            //.parallel()
            //.mapToLong(seed -> FromSeedToLocation(seed))
            
        long lowest = Long.MAX_VALUE;

       
        for (int i = 0; i < locations.length - 1; i += 2) {
            long start = locations[i];
            long end = locations[i + 1];

            long[] temp = LongStream.range(start, start + end)
                .parallel()
                .map(DayFive::FromSeedToLocation)
                .toArray();
            Arrays.sort(temp);

            lowest = temp[0] > lowest ? lowest : temp[0]; 
        }
        //Arrays.sort(locations);
        //Arrays.sort(part2);

        System.out.println("Part 1: " + lowest);

    }

    private static long FromSeedToLocation (long seed) {
        
        for (int i = 3 ; i < fileLength; i++) {

            String[] currentLine = file.get(i).split("\\D+");
            boolean reachedEOF = i >= fileLength - 1;

            while (currentLine.length > 1 && !reachedEOF) {            
                long destinationStart = Long.parseLong(currentLine[0]);
                long sourceStart = Long.parseLong(currentLine[1]);
                long range = Long.parseLong(currentLine[2]);
                final long seedCopy = seed;


                /* int[] destinationRange = IntStream.range(destinationStart, destinationStart + range)
                    .filter(num -> num == seedCopy)
                    .toArray(); */


                long[] sourceRange = LongStream.range(sourceStart, sourceStart + range)
                    .filter(num -> num == seedCopy)
                    .toArray();
                if (sourceRange.length > 0) {
                    boolean some = true;
                    seed = destinationStart + ( seed - sourceStart);
                    while (some) {
                        i++;
                        
                        String[] nextLine = i <= fileLength - 1 ? file.get(i).split("\\D+") : "".split("\\D+");
                        
                        if (i >= fileLength || nextLine.length < 2) {
                            some = false;
                            break;
                        }
                    }
                    break;
                }

                if (i + 1 == fileLength) { reachedEOF = true; break;}

                i++;

                currentLine = file.get(i).split("\\D+");


                if (currentLine.length <= 1) {
                    
                    break;
                }
            } 
            if(reachedEOF) {
                return seed;
            } 
        }
        return seed;
    }

}
